File: 337-house-robber-iii/337-house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        unordered_map<TreeNode*, int> dp;
        return solve(root, dp);
    }
private:
    int solve(TreeNode* root, unordered_map<TreeNode*, int>& dp) {
        if(root == nullptr)
            return 0;
        
        if(dp.find(root) != dp.end()) return dp[root];
        
        int notPick = solve(root -> left, dp) + solve(root -> right, dp);
        
        int pick = root -> val;
        
        if(root -> left)
            pick += solve(root -> left -> left, dp) + solve(root -> left -> right, dp);
        
        if(root -> right)
            pick += solve(root -> right -> left, dp) + solve(root -> right -> right, dp);
        
        return dp[root] = max(pick, notPick);
    }
};

File: 337-house-robber-iii/337-house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def solve(root, dp):
            if root == None:
                return 0
            
            if root in dp:
                return dp[root]
            
            pick = solve(root.left, dp) + solve(root.right, dp)
            
            notPick = root.val
            
            if root.left:
                notPick += solve(root.left.left, dp) + solve(root.left.right, dp)
            
            if root.right:
                notPick += solve(root.right.left, dp) + solve(root.right.right, dp)
            
            dp[root] = max(pick, notPick)
            return dp[root]
        
        dp = dict()
        return solve(root, dp)
            